controllers/openstackdataplanerole_controller.go

```diff
@@ -184,26 +183,49 @@ func (r *OpenStackDataPlaneRoleReconciler) Reconcile(ctx context.Context, req ct
 	if err != nil {
 		return ctrl.Result{}, err
 	}
-	nodeNames := ""
-	for _, node := range nodes.Items {
-		if node.Spec.Role != instance.Name {
-			err = fmt.Errorf("node %s: node.Role does not match with node.Label", node.Name)
-			return ctrl.Result{}, err
-		}
-		nodeNames = nodeNames + node.Name + ","
-	}
-	r.Log.Info("Role", "Nodes", nodeNames, "Role.Namespace", instance.Namespace, "Role.Name", instance.Name)
+
+	// Order the nodes based on Name
+	sort.SliceStable(nodes.Items, func(i, j int) bool {
+		return nodes.Items[i].Name < nodes.Items[j].Name
+	})
+
+	// Validate NodeSpecs
 	err = instance.Validate(nodes.Items)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 
-	roleConfigMap, err := r.GenerateInventory(ctx, instance, nodes.Items)
+	// Ensure IPSets Required for Nodes
+	allIPSets, isReady, err := deployment.EnsureIPSets(ctx, helper, instance, nodes)
+	if err != nil || !isReady {
+		return ctrl.Result{}, err
+	}
+
+	// Ensure DNSData Required for Nodes
+	dnsAddresses, isReady, err := deployment.EnsureDNSData(ctx, helper,
+		instance, nodes, allIPSets)
+	if err != nil || !isReady {
+		return ctrl.Result{}, err
+	}
+
+	// Reconcile BaremetalSet if required
+	if len(instance.Spec.BaremetalSetTemplate.BaremetalHosts) > 0 {
+		isReady, err := deployment.DeployBaremetalSet(ctx, helper, instance,
+			nodes, allIPSets, dnsAddresses)
+		if err != nil || !isReady {
+			return ctrl.Result{}, err
+		}
+	}
```

Comment:
Maybe as a follow-up later on we could add some condition updates here to let the user know what is going on while these are processing (and when they've finished).  Otherwise the whole `SetupReadyCondition` is masking a lot of intermediate steps that themselves might take some time (especially the BaremetalSet).

Reply:
We do update RoleBareMetalProvisionReadyCondition, RoleDNSDataReadyCondition, RoleIPReservationReadyConditions before SetupReadyCondition. Is that what you mean or  updating waiting messages for these additional conditions as it's waiting for them for reconciliation to proceed further?

Reply:
Sorry, I missed that you're setting the conditions in a different file outside of the controller.  Looks good as-is.